drafts/AENS.md

```diff
@@ -0,0 +1,414 @@
+
+
+### Hashing
+
+We are going to adopt the `NameHash` as defined by the ENS in 
+[EIP-137](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-137.md#namehash-algorithm)
+using SHA256 instead.
+
+```
+def namehash(name):
+  if name == '':
+    return '\0' * 32
+  else:
+    label, _, remainder = name.partition('.')
+    return sha256(namehash(remainder) + sha256(label))
+```
+
+Names are generally only referred to in hashed form.
+
+### Protocol
+
+```
+                  expire
+      unclaimed <-------- revoked
+          | ^              ^^
+          | |              || expire
+          | | expire       ||
+pre-claim | |              ||  _
+          | |       revoke || | | transfer
+          v |              || | v
+     pre-claimed -------> claimed
+                  claim    | ^   
+                           | |
+                            -
+                          update
+```
+
+Note that `expire` is not an explicit message that is part
+of the protocol.
+
+
+#### Pre-claim
+
+Send a pre-claim transaction containing a hash commitment.
+
+```
+ ------------ ----
+| hash       | 32 |
+ ------------ ----
+```
+
+
+
+#### Claim
+
+```
+ ------------ ----
+| name       | 63 |
+ ------------ ----
+| fee        | 32 |
+ ------------ ----
+```
+
+Flow for a user:
+
+1. (optional) wait `n` blocks, s.t. that the block including the `pre-claim` cannot be reversed whp 
+2. send `claim` transaction to reveal name and pay the associated fee
+
+If the time delta of 1 and 2 is bigger than e.g. 48 hours then the 
+`pre-claim` expires again.
+
+The `claim` transaction MUST be signed by the same private key as the 
+`pre-claim` transaction.
+
+
+#### Update
+
+```
+ ------------ ----
+| hash       | 32 |
+ ------------ ----
+| expire_by  | 8  |
+ ------------ ----
+| ttl        | 8  |
+ ------------ ----
+| pointers   |    |
+ ------------ ----
+```
+
+The `update` transaction MUST be signed by the owner
+of the name entry to be updated.
+
+The `expire_by` MUST NOT be more than 1 year into the
+future.
+
+
+#### Transfer
+
+```
+ ------------ ----
+| target     | 32 |
+ ------------ ----
+| hash       | 32 |
+ ------------ ----
+```
+
+The `transfer` transaction MUST be signed by the owner
+of the name entry to be transfered.
+
+
+### Revoke
+
+```
+ ------------ ----
+| hash       | 32 |
+ ------------ ----
+```
+
+The revoke transaction MUST be signed by the owner
+of a name entry.
+
+After the `revoke` transaction has been included in the chain,
+the name enters the `revoked` state. After a fixed timeout of
+2 weeks, the name will be available for claiming again.
+
+
+## Storage
+
+We are going to store the AENS entries in an ESMT and use the
+hash of a name, as defined above, as a pointer to the entry data.
+If there's no entry for a given hash then that pointer will point
+to an empty hash or to the hash of that entry data otherwise.
+
+
+## Launch
+
+Giving users plenty time to prepare for a launch of the naming
```

Comment:
Is the .test namespace intended to be available as soon as the feature is released and other top level domains only available after launch?
This would make it possible to write clients and test the feature before the launch.

Reply:
Yes, the `.test` namespace should be available on the testnet, as soon as possible.